File: src/main/swift/g0001_0100/s0096_unique_binary_search_trees/Solution.swift
// #Medium #Dynamic_Programming #Math #Tree #Binary_Tree #Binary_Search_Tree
// #Dynamic_Programming_I_Day_11 #Big_O_Time_O(n)_Space_O(1)
// #2024_06_25_Time_0_ms_(100.00%)_Space_15.3_MB_(56.25%)

class Solution {
    func numTrees(_ n: Int) -> Int {
        var result: Int64 = 1
        for i in 0..<n {
            result *= Int64(2 * n - i)
            result /= Int64(i + 1)
        }
        result /= Int64(n + 1)
        return Int(result)
    }
}

File: src/main/js/g0001_0100/s0096_unique_binary_search_trees/solution.js
// #Medium #Dynamic_Programming #Math #Tree #Binary_Tree #Binary_Search_Tree
// #Dynamic_Programming_I_Day_11 #Big_O_Time_O(n)_Space_O(1)
// #2024_12_12_Time_0_ms_(100.00%)_Space_48.8_MB_(65.65%)

/**
 * @param {number} n
 * @return {number}
 */
var numTrees = function (n) {
    let result = 1

    for (let i = 0; i < n; i++) {
        result *= 2 * n - i
        result /= i + 1
    }

    result /= n + 1
    return Math.floor(result)
};

export { numTrees }
